prover/src/lib.rs

```diff
@@ -79,17 +71,53 @@ pub fn prove(
 
 struct ExecutionProver {
     options: ProofOptions,
-    num_stack_inputs: usize,
-    num_stack_outputs: usize,
+    stack_inputs: Vec<Felt>,
+    outputs: ProgramOutputs,
 }
 
 impl ExecutionProver {
-    pub fn new(options: ProofOptions, num_stack_inputs: usize, num_stack_outputs: usize) -> Self {
+    pub fn new(options: ProofOptions, stack_inputs: Vec<Felt>, outputs: ProgramOutputs) -> Self {
         Self {
             options,
-            num_stack_inputs,
-            num_stack_outputs,
+            stack_inputs,
+            outputs,
+        }
+    }
+
+    // HELPER FUNCTIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Validates the stack inputs against the provided execution trace and returns true if valid.
+    fn validate_inputs(&self, trace: &ExecutionTrace) -> bool {
+        let mut valid = true;
+        for (i, input) in trace
+            .init_stack_state()
+            .iter()
+            .take(self.stack_inputs.len())
+            .enumerate()
+        {
+            if *input != self.stack_inputs[i] {
+                valid = false;
+            }
         }
+
+        valid
+    }
```

Comment:
There probably should be a `break;` after `valid = false;` here and in the `validate_ouputs()` method below.

nit: since these return booleans, I'd prefer to name them something like `are_inputs_valid()` and `are_outputs_valid()`.

Also, curious why you chose to index into `stack_inputs` rather than doing something like:
```Rust
for (t_input, p_input) in trace.init_stack_state().iter().zip(self.stack_inputs()) {
```

Reply:
Yep, happy to add the break and change the names. The reason I didn't use zip is that there's no guarantee they're the same length. `stack_inputs` could be shorter or longer than `init_stack_state`. If I did it the way you proposed we'd have to have more length checks, so I thought it would be more complex and less readable.

Reply:
I might be wrong, but I thought `zip` would produce elements only until both iterators return `Some(...)`. That is, as soon as one of the iterators stops, `zip`'ed iterator would stop as well.

Reply:
Ah you're right! I thought I'd read it the other way, but I clearly couldn't have, because the docs agree with you :) Yep, `zip` would be more convenient.